.NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs

```diff
@@ -157,6 +93,221 @@ public DateTimeParseResult Parse(ExtractResult er, DateObject refDate)
+                DateObject futureEnd = DateObject.MinValue;
+                DateObject pastBegin = DateObject.MinValue;
+                DateObject pastEnd = DateObject.MinValue;
+                var isSpecificDate = false;
+                var isStartByWeek = false;
+                var isEndByWeek = false;
+
+                var startResolution = ParseSingleTimePoint(match.Groups["start"].Value.Trim(), referenceDate);
+
+                if (startResolution.Success)
+                {
+                    futureBegin = (DateObject)startResolution.FutureValue;
+                    pastBegin = (DateObject)startResolution.PastValue;
+                    isSpecificDate = true;
+                }
+                else
+                {
+                    startResolution = ParseBaseDatePeriod(match.Groups["start"].Value.Trim(), referenceDate);
+
+                    if (startResolution.Success)
+                    {
+                        futureBegin = ((Tuple<DateObject, DateObject>)startResolution.FutureValue).Item1;
+                        pastBegin = ((Tuple<DateObject, DateObject>)startResolution.PastValue).Item1;
+                        
+                        if (startResolution.Timex.Contains("-W"))
+                        {
+                            isStartByWeek = true;
+                        }
+                    }
+                }
+
+                if (startResolution.Success)
+                {
+                    var endResolution = ParseSingleTimePoint(match.Groups["end"].Value.Trim(), referenceDate);
+
+                    if (endResolution.Success)
+                    {
+                        futureEnd = (DateObject)endResolution.FutureValue;
+                        pastEnd = (DateObject)endResolution.PastValue;
+                        isSpecificDate = true;
+                    }
+                    else
+                    {
+                        endResolution = ParseBaseDatePeriod(match.Groups["end"].Value.Trim(), referenceDate);
+
+                        if (endResolution.Success)
+                        {
+                            futureEnd = ((Tuple<DateObject, DateObject>)endResolution.FutureValue).Item1;
+                            pastEnd = ((Tuple<DateObject, DateObject>)endResolution.PastValue).Item1;
+                            if (endResolution.Timex.Contains("-W"))
+                            {
+                                isEndByWeek = true;
+                            }
+                        }
+                    }
+
+                    if (endResolution.Success)
+                    {
+                        if (futureBegin > futureEnd)
+                        {
+                            futureBegin = pastBegin;
+                        }
+
+                        if (pastEnd < pastBegin)
+                        {
+                            pastEnd = futureEnd;
+                        }
+
+                        if (isSpecificDate)
```

Comment:
Couldn't the timex generation be a method somewhere for re-use?
It would also make future maintenance easier and avoid introducing inconsistencies throughout the codebase.

Reply:
Fixed.